src/preamble.js

```diff
@@ -164,42 +171,41 @@ function ccall (ident, returnType, argTypes, args, opts) {
     }
   }
   var ret = func.apply(null, cArgs);
-#if ASSERTIONS
 #if EMTERPRETIFY_ASYNC
-  if ((!opts || !opts.async) && typeof EmterpreterAsync === 'object') {
-    assert(!EmterpreterAsync.state, 'cannot start async op with normal JS calling ccall');
-  }
-  if (opts && opts.async) assert(!returnType, 'async ccalls cannot return values');
+  if (typeof EmterpreterAsync === 'object' && EmterpreterAsync.state) {
+#if ASSERTIONS
+    assert(opts && opts.async, 'The call to ' + ident + ' is running asynchronously. If this was intended, add the async option to the ccall/cwrap call.');
 #endif
+    return new Promise(function(resolve) {
+      EmterpreterAsync.restartFunc = func;
```

Comment:
please add something like
```
#if ASSERTIONS
  assert(!EmterpreterAsync.restartFunc, 'Cannot have multiple async ccalls in flight at once');
#endif
```

Reply:
If I'm not mistaken, this should already be caught by https://github.com/jscissr/emscripten/blob/55783d98ca68caa540509d8ac372103543ef24a4/src/library_async.js#L355 (when an async call is active, `EmterpreterAsync.state` is 1).

Reply:
Yeah, this should not be a reachable assertion in principle, I agree, but I think it's still useful: it makes reading the code easier (at least for me), and also it protects us against future refactoring of the code, say if that other assertion were removed for some (good or bad) reason.